src/economy/economy.cc

```diff
@@ -470,13 +483,16 @@ bool Economy::needs_ware_or_worker(DescriptionIndex const ware_or_worker_type) c
 			}
 		}
 		return true;
-
-		// we have target quantity set to 0, we need to check if there is an open request
 	} else {
-		for (const Request* temp_req : requests_) {
-			const Request& req = *temp_req;
-
-			if (req.get_type() == type_ && req.get_index() == ware_or_worker_type) {
+		// Target quantity is set to 0, we need to check if there is an open request.
+		// For soldier requests, do not recruit new rookies if only heroes are needed.
+		const SoldierDescr* sd =
+		   type_ == wwWORKER ? dynamic_cast<const SoldierDescr*>(
+		                          owner().egbase().tribes().get_worker_descr(ware_or_worker_type)) :
```

Comment:
You can compare the index to `owner().tribe().soldier()` and lose the dynamic cast

Reply:
Okay will do so